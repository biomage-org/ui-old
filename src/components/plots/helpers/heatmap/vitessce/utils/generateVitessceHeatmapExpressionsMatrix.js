import _ from 'lodash';
import { convertRange } from 'utils/plotUtils';

const scaledTo255 = (rowOfExpressions, min, max) => (
  rowOfExpressions.map((value) => convertRange(value, [min, max], [0, 255]))
);

const generateVitessceHeatmapExpressionsMatrix = (cellOrder, geneOrder, expressionMatrix) => {
  const geneExpressionsDataMatrix = [];

  geneOrder.forEach((gene) => {
    if (!expressionMatrix.geneIsLoaded(gene)) return;
    const truncatedExpression = expressionMatrix.getTruncatedExpression(gene, cellOrder);

    const { truncatedMin, truncatedMax } = expressionMatrix.getStats(gene);

    const scaledGeneExpressions = scaledTo255(truncatedExpression, truncatedMin, truncatedMax);

    geneExpressionsDataMatrix.push(scaledGeneExpressions);
  });

  return _.flatten(_.unzip(geneExpressionsDataMatrix));
};

export default generateVitessceHeatmapExpressionsMatrix;
