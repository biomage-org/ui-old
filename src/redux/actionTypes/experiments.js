const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Experiments loaded
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Example experiments loaded
 */
const EXPERIMENTS_EXAMPLES_LOADED = `${EXPERIMENTS}/examplesLoaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

const EXPERIMENTS_SAVED = `${EXPERIMENTS}/saved`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

/**
 * Set active project.
 */
const EXPERIMENTS_SET_ACTIVE = `${EXPERIMENTS}/setActive`;

/**
 * Add project metadata.
 */
const EXPERIMENTS_METADATA_CREATE = `${EXPERIMENTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const EXPERIMENTS_METADATA_RENAME = `${EXPERIMENTS}/metadataRename`;

/**
 * Delete project metadta.
 */
const EXPERIMENTS_METADATA_DELETE = `${EXPERIMENTS}/metadataDelete`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_EXAMPLES_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_SWITCH,
  EXPERIMENTS_SET_ACTIVE,
  EXPERIMENTS_METADATA_CREATE,
  EXPERIMENTS_METADATA_RENAME,
  EXPERIMENTS_METADATA_DELETE,
};
