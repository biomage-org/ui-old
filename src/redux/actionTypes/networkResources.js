const NETWORK_RESOURCES = 'networkResources';

/**
 * Loads name environment for the current request.
 */
const NETWORK_RESOURCES_DEPLOYMENT_INFO_LOADED = `${NETWORK_RESOURCES}/deploymentInfoLoaded`;

export {
  // eslint-disable-next-line import/prefer-default-export
  NETWORK_RESOURCES_DEPLOYMENT_INFO_LOADED,
};
