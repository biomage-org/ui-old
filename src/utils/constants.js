const api = {
  V1: 'v1',
  V2: 'v2',
};

const modules = {
  DATA_MANAGEMENT: 'DATA_MANAGEMENT',
  DATA_PROCESSING: 'DATA_PROCESSING',
  DATA_EXPLORATION: 'DATA_EXPLORATION',
  PLOTS_AND_TABLES: 'PLOTS_AND_TABLES',
  SETTINGS: 'SETTINGS',
  DEFAULT: '',
};

const sampleTech = {
  '10X': '10x',
  RHAPSODY: 'rhapsody',
};

const plotTypes = {
  CONTINUOUS_EMBEDDING: 'ContinuousEmbedding',
  CATEGORICAL_EMBEDDING: 'CategoricalEmbedding',
  HEATMAP: 'Heatmap',
  MARKER_HEATMAP: 'MarkerHeatmap',
  VOLCANO_PLOT: 'VolcanoPlot',
  FREQUENCY_PLOT: 'FrequencyPlot',
  VIOLIN_PLOT: 'violin',
  DOT_PLOT: 'DotPlot',
  TRAJECTORY_ANALYSIS: 'TrajectoryAnalysis',
  NORMALIZED_EXPRESSION_MATRIX: 'NormalizedExpressionMatrix',
  IMG_PLOT: 'ImgPlot',
};

const plotNames = {
  CONTINUOUS_EMBEDDING: 'Continuous Embedding',
  CATEGORICAL_EMBEDDING: 'Categorical Embedding',
  HEATMAP: 'Custom Heatmap',
  MARKER_HEATMAP: 'Marker Heatmap',
  VOLCANO_PLOT: 'Volcano Plot',
  FREQUENCY_PLOT: 'Frequency Plot',
  VIOLIN_PLOT: 'Violin Plot',
  DOT_PLOT: 'Dot Plot',
  TRAJECTORY_ANALYSIS: 'Trajectory Analysis',
  NORMALIZED_EXPRESSION_MATRIX: 'Normalized Expression Matrix',
  IMG_PLOT: 'Image plot'
};

const layout = {
  PANEL_HEADING_HEIGHT: 30,
  PANEL_PADDING: 10,
};

export {
  api,
  modules,
  sampleTech,
  plotTypes,
  plotNames,
  layout,
};
