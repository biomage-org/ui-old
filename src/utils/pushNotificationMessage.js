import { message } from 'antd';

const pushNotificationMessage = (type, text, duration = 4) => {
  switch (type) {
    case 'success':
      message.success(text, duration);
      break;
    case 'error':
      message.error(text, duration);
      break;
    case 'info':
      message.info(text, duration);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, duration);
      break;
    case 'loading':
      message.loading(text, duration);
      break;
    default:
      message.info(text, duration);
      break;
  }
};

export default pushNotificationMessage;
