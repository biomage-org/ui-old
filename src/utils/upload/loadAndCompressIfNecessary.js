import { gzip } from 'fflate';
import readFileToBuffer from 'utils/upload/readFileToBuffer';

const compress = (buffer) => new Promise((resolve, reject) => {
  gzip(buffer, {}, (error, compressed) => {
    if (error) {
      reject(new Error('error'));
    }
    resolve(Buffer.from((compressed)));
  });
});

// eslint-disable-next-line arrow-body-style
const loadAndCompressIfNecessary = (file, onCompression = () => ({})) => {
  return readFileToBuffer(file.fileObject)
    .then((buffer) => {
      if (file.compressed) {
        return buffer;
      }
      onCompression();
      return compress(buffer);
    });
};

export default loadAndCompressIfNecessary;
