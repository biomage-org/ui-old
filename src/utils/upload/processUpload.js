/* eslint-disable no-param-reassign */
import _ from 'lodash';

import axios from 'axios';

import {
  createSamples, createSampleFile, updateSampleFileUpload, validateSamples,
} from 'redux/actions/samples';

import UploadStatus from 'utils/upload/UploadStatus';
import loadAndCompressIfNecessary from 'utils/upload/loadAndCompressIfNecessary';
import { inspectFile, Verdict } from 'utils/upload/fileInspector';
import getFileTypeV2 from 'utils/getFileTypeV2';
import { sampleTech } from 'utils/constants';

const MAX_RETRIES = 2;

const putInS3 = async (loadedFileData, signedUrl, onUploadProgress, currentRetry = 0) => {
  try {
    await axios.request({
      method: 'put',
      url: signedUrl,
      data: loadedFileData,
      headers: {
        'Content-Type': 'application/octet-stream',
      },
      onUploadProgress,
    });
  } catch (e) {
    if (currentRetry < MAX_RETRIES) {
      await putInS3(loadedFileData, signedUrl, onUploadProgress, currentRetry + 1);
    }

    throw e;
  }
};

const createAndUploadSingleFile = async (
  file, experimentId, sampleId, dispatch, selectedTech,
) => {
  let loadedFile = null;

  const fileType = getFileTypeV2(file.fileObject.name, file.fileObject.type);

  try {
    loadedFile = await loadAndCompressIfNecessary(file, () => {
      dispatch(updateSampleFileUpload(experimentId, sampleId, fileType, UploadStatus.COMPRESSING));
    });
  } catch (e) {
    const fileErrorStatus = e.message === 'aborted' ? UploadStatus.FILE_READ_ABORTED : UploadStatus.FILE_READ_ERROR;

    dispatch(updateSampleFileUpload(experimentId, sampleId, fileType, fileErrorStatus));
    return;
  }

  try {
    const metadata = getMetadata(file, selectedTech);

    const signedUrl = await dispatch(
      createSampleFile(
        experimentId,
        sampleId,
        fileType,
        file.size,
        metadata,
        file,
      ),
    );

    await putInS3(loadedFile, signedUrl, (progress) => {
      const percentProgress = Math.round((progress.loaded / progress.total) * 100);

      dispatch(
        updateSampleFileUpload(
          experimentId, sampleId, fileType, UploadStatus.UPLOADING, percentProgress ?? 0,
        ),
      );
    });
  } catch (e) {
    dispatch(updateSampleFileUpload(experimentId, sampleId, fileType, UploadStatus.UPLOAD_ERROR));
    return;
  }

  await dispatch(updateSampleFileUpload(experimentId, sampleId, fileType, UploadStatus.UPLOADED));
};

const getMetadata = (file, selectedTech) => {
  const metadata = {};
  if (selectedTech === sampleTech['10X']) {
    if (file.name.includes('genes')) {
      metadata.cellranger_version = 'v2';
    } else if (file.name.includes('features')) {
      metadata.cellranger_version = 'v3';
    }
  }
  return metadata;
};

const processUpload = async (filesList, technology, samples, experimentId, dispatch) => {
  // First use map to make it easy to add files in the already existing sample entry
  const samplesMap = filesList.reduce((acc, file) => {
    const pathToArray = file.name.trim().replace(/[\s]{2,}/ig, ' ').split('/');

    const sampleName = pathToArray[0];
    const fileName = _.last(pathToArray);

    // Update the file name so that instead of being saved as
    // e.g. WT13/matrix.tsv.gz, we save it as matrix.tsv.gz
    file.name = fileName;

    const sampleUuid = Object.values(samples).filter(
      (s) => s.name === sampleName
        && s.experimentId === experimentId,
    )[0]?.uuid;

    return {
      ...acc,
      [sampleName]: {
        ...acc[sampleName],
        uuid: sampleUuid,
        files: {
          ...acc[sampleName]?.files,
          [fileName]: file,
        },
      },
    };
  }, {});

  const validSamplesList = await dispatch(validateSamples(experimentId, samplesMap, technology));

  // If none of the files are in valid format, return
  if (validSamplesList.length === 0) return;

  // Sort alphabetically
  validSamplesList.sort(([oneName], [otherName]) => oneName.localeCompare(otherName));

  try {
    const sampleIdsByName = await dispatch(
      createSamples(
        experimentId,
        validSamplesList,
        technology,
      ),
    );

    const promises = [];

    validSamplesList.forEach(([name, sample]) => {
      Object.values(sample.files).forEach((file) => {
        promises.push(
          async () => await createAndUploadSingleFile(
            file,
            experimentId,
            sampleIdsByName[name],
            dispatch,
            technology,
          ),
        );
      });
    });

    // 5 at a time
    const chunkedPromises = _.chunk(promises, promises.length);

    // eslint-disable-next-line no-restricted-syntax
    for await (const promisesChunk of chunkedPromises) {
      await Promise.all(promisesChunk.map((promise) => promise()));
    }
  } catch (e) {
    // Ignore the error, if createSamples fails we throw to
    // avoid attempting to upload any of these broken samples
  }
};

/**
   * This function converts an uploaded File object into a file record that will be inserted under
   * samples[files] in the redux store.
   * @param {File} fileObject the File object that is uploaded to the .
   * @param {string} technology the chosen technology that outputs this file. Used for verification.
   * @returns {object} fileRecord object that will be associated with a sample.
   */
const fileObjectToFileRecord = async (fileObject, technology) => {
  // This is the first stage in uploading a file.

  // if the file has a path, trim to just the file and its folder.
  // otherwise simply use its name
  const filename = (fileObject.path)
    ? _.takeRight(fileObject.path.split('/'), 2).join('/')
    : fileObject.name;

  const verdict = await inspectFile(fileObject, technology);

  let error = '';
  if (verdict === Verdict.INVALID_NAME) {
    error = 'Invalid file name.';
  } else if (verdict === Verdict.INVALID_FORMAT) {
    error = 'Invalid file format.';
  }

  return {
    name: filename,
    fileObject,
    size: fileObject.size,
    path: fileObject.path,
    upload: {
      status: UploadStatus.UPLOADING,
      progress: 0,
    },
    valid: !error,
    errors: error,
    compressed: verdict === Verdict.VALID_ZIPPED,
  };
};

export {
  fileObjectToFileRecord,
  createAndUploadSingleFile,
};

export default processUpload;
