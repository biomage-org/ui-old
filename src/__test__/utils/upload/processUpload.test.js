import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import waitForActions from 'redux-mock-store-await-actions';
import axios from 'axios';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 } from 'uuid';

import { SAMPLES_FILE_UPDATE } from 'redux/actionTypes/samples';
import initialSampleState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import UploadStatus from 'utils/upload/UploadStatus';
import { waitFor } from '@testing-library/dom';

import processUpload from 'utils/upload/processUpload';

import loadAndCompressIfNecessary from 'utils/upload/loadAndCompressIfNecessary';
import validate from 'utils/upload/validate10x';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import { sampleTech } from 'utils/constants';

enableFetchMocks();

const getValidFiles = (cellrangerVersion, sampleName = 'WT13') => {
  const filename = cellrangerVersion === 'v2' ? 'genes.tsv.gz' : 'features.tsv.gz';

  return ([
    {
      name: `${sampleName}/${filename}`,
      fileObject: {
        name: filename,
        path: `/${sampleName}/${filename}`,
        type: 'application/gzip',
        size: 100,
      },
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed: true,
      valid: true,
    },
    {
      name: `${sampleName}/barcodes.tsv.gz`,
      fileObject: {
        name: 'barcodes.tsv.gz',
        path: `/${sampleName}/barcodes.tsv.gz`,
        type: 'application/gzip',
        size: 100,
      },
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed: true,
      valid: true,
    },
    {
      name: `${sampleName}/matrix.mtx.gz`,
      fileObject: {
        name: 'matrix.mtx.gz',
        path: `/${sampleName}/matrix.mtx.gz`,
        type: 'application/gzip',
        size: 100,
      },
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed: true,
      valid: true,
    },
  ]);
};

const sampleType = sampleTech['10X'];
const mockSampleUuid = 'sample-uuid';
const mockExperimentId = 'project-uuid';
const sampleName = 'mockSampleName';

const mockUnrelatedSampleUuid = 'unrelated-sample-uuid';
const mockUnrelatedExperimentId = 'unrelated-experiment-id';

const initialState = {
  experiments: {
    ...initialExperimentState,
    [mockExperimentId]: {
      ...experimentTemplate,
      id: mockExperimentId,
      sampleIds: [mockSampleUuid],
    },
  },
  samples: {
    ...initialSampleState,
    ids: [mockSampleUuid],
    meta: {
      loading: true,
      error: false,
    },
    [mockSampleUuid]: {
      ...sampleTemplate,
      uuid: mockSampleUuid,
      name: sampleName,
      experimentId: mockExperimentId,
    },
    [mockUnrelatedSampleUuid]: {
      ...sampleTemplate,
      uuid: mockUnrelatedSampleUuid,
      name: sampleName,
      experimentId: mockUnrelatedExperimentId,
    },
  },
};
// Based on https://stackoverflow.com/a/51045733
const flushPromises = () => new Promise(setImmediate);
const mockStore = configureMockStore([thunk]);

jest.mock('utils/upload/loadAndCompressIfNecessary',
  () => jest.fn().mockImplementation(
    (file) => {
      if (!file.valid) {
        return Promise.reject(new Error('error'));
      }
      return Promise.resolve('loadedGzippedFile');
    },
  ));

jest.mock('uuid');

jest.mock('axios', () => ({
  request: jest.fn(),
}));

jest.mock('utils/pushNotificationMessage');

jest.mock('utils/upload/validate10x');

let store = null;

describe('processUpload', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResponse(JSON.stringify('theSignedUrl'), { status: 200 });

    store = mockStore(initialState);

    v4.mockReturnValue('sample-uuid');
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v3', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve('Resolved');
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toEqual('loadedGzippedFile');
    expect(mockAxiosCalls[1].data).toEqual('loadedGzippedFile');
    expect(mockAxiosCalls[2].data).toEqual('loadedGzippedFile');

    // Wait until all put promises are resolved
    await flushPromises();

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // Calls to loadAndCompressIfNecessary are done correctly
    expect(loadAndCompressIfNecessary.mock.calls.map((calls) => calls[0])).toMatchSnapshot();

    // If we trigger onCompression then an action to update uploadStatus to COMPRESSING is made
    const onCompression = loadAndCompressIfNecessary.mock.calls[0][1];
    onCompression();

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.COMPRESSING } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: 20, total: 40 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v2', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve('Resolved');
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    await processUpload(
      getValidFiles('v2'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toEqual('loadedGzippedFile');
    expect(mockAxiosCalls[1].data).toEqual('loadedGzippedFile');
    expect(mockAxiosCalls[2].data).toEqual('loadedGzippedFile');

    // Wait until all put promises are resolved
    await flushPromises();

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // Calls to loadAndCompressIfNecessary are done correctly
    expect(loadAndCompressIfNecessary.mock.calls.map((calls) => calls[0])).toMatchSnapshot();

    // If we trigger onCompression then an action to update uploadStatus to COMPRESSING is made
    const onCompression = loadAndCompressIfNecessary.mock.calls[0][1];
    onCompression();

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.COMPRESSING } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: 20, total: 40 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );
  });

  it('Uploads following alphabetical order for new samples', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve('Resolved');
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    const firstSampleFiles = getValidFiles('v3', 'aaaFirstSample');
    const secondSampleFiles = getValidFiles('v3', 'bbbSecondSample');
    const thirdSampleFiles = getValidFiles('v3', 'cccThirdSample');

    // Scramble the order in which they are sent
    const allFiles = [thirdSampleFiles, firstSampleFiles, secondSampleFiles].flat();

    const mockUuids = ['uuid3', 'uuid2', 'uuid1'];

    v4
      .mockReset()
      .mockImplementation(() => mockUuids.pop());

    await processUpload(
      allFiles,
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait until all put promises are resolved
    await flushPromises();

    const first3Calls = fetchMock.mock.calls.slice(0, 3);

    expect(first3Calls).toEqual([
      [
        expect.stringContaining('uuid1'),
        // eslint-disable-next-line no-useless-escape
        expect.objectContaining({ method: 'POST', body: expect.stringContaining('\"name\":\"aaaFirstSample\"') }),
      ],
      [
        expect.stringContaining('uuid2'),
        // eslint-disable-next-line no-useless-escape
        expect.objectContaining({ method: 'POST', body: expect.stringContaining('\"name\":\"bbbSecondSample\"') }),
      ],
      [
        expect.stringContaining('uuid3'),
        // eslint-disable-next-line no-useless-escape
        expect.objectContaining({ method: 'POST', body: expect.stringContaining('\"name\":\"cccThirdSample\"') }),
      ],
    ]);
  });

  it('Updates redux correctly when there are file load and compress errors', async () => {
    const invalidFiles = getValidFiles('v3').map((file) => ({ ...file, valid: false }));

    await processUpload(
      invalidFiles,
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.FILE_READ_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );

    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);

    const uploadingFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const errorFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.FILE_READ_ERROR,
    );
    const uploadedFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );
    // There are 3 files actions with status uploading
    expect(uploadingFileProperties.length).toEqual(3);
    // There are 3 files actions with status upload error
    expect(errorFileProperties.length).toEqual(3);
    // There are no file actions with status successfully uploaded
    expect(uploadedFileProperties.length).toEqual(0);
  });

  it('Updates redux correctly when there are file upload errors', async () => {
    const mockAxiosCalls = [];

    const uploadError = (params) => {
      mockAxiosCalls.push(params);
      return Promise.reject(new Error('Error'));
    };

    axios.request.mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError);

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOAD_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );

    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);

    const uploadingFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );

    const errorFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOAD_ERROR,
    );

    const uploadedFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingFileProperties.length).toEqual(3);
    // There are 3 files actions with status upload error
    expect(errorFileProperties.length).toEqual(3);
    // There are no file actions with status successfully uploaded
    expect(uploadedFileProperties.length).toEqual(0);
  });

  it('Should not upload files if there are errors creating samples in the api', async () => {
    fetchMock.mockReject(new Error('Error'));

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(axios.request).not.toHaveBeenCalled();
    });
  });

  it('Should not upload sample and show notification if uploaded sample is invalid', async () => {
    validate.mockImplementationOnce(
      () => { throw new Error('Some file error'); },
    );

    await processUpload(
      getValidFiles('v2'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
      expect(axios.request).not.toHaveBeenCalled();
    });
  });
});
