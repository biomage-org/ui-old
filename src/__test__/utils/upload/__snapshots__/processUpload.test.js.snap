// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v2 1`] = `
[
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "genes.tsv.gz",
      "path": "/WT13/genes.tsv.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "genes.tsv.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "barcodes.tsv.gz",
      "path": "/WT13/barcodes.tsv.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "barcodes.tsv.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "matrix.mtx.gz",
      "path": "/WT13/matrix.mtx.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "matrix.mtx.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v2 2`] = `
[
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v3 1`] = `
[
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "features.tsv.gz",
      "path": "/WT13/features.tsv.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "features.tsv.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "barcodes.tsv.gz",
      "path": "/WT13/barcodes.tsv.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "barcodes.tsv.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
  {
    "compressed": true,
    "errors": "",
    "fileObject": {
      "name": "matrix.mtx.gz",
      "path": "/WT13/matrix.mtx.gz",
      "size": 100,
      "type": "application/gzip",
    },
    "name": "matrix.mtx.gz",
    "upload": {
      "status": "uploading",
    },
    "valid": true,
  },
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v3 2`] = `
[
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  {
    "data": "loadedGzippedFile",
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
]
`;
